src/app/ReadClient.cpp

```diff
@@ -720,5 +753,42 @@ CHIP_ERROR ReadClient::SendSubscribeRequest(ReadPrepareParams & aReadPreparePara
     return CHIP_NO_ERROR;
 }
 
-}; // namespace app
-}; // namespace chip
+void ReadClient::OnResubscribeTimerCallback(System::Layer * apSystemLayer, void * apAppState)
+{
+    ReadClient * const client = reinterpret_cast<ReadClient *>(apAppState);
+    if (client != nullptr)
+    {
+        client->SendSubscribeRequest(client->mReadPrepareParams);
+        client->mNumRetries++;
+    }
+    else
+    {
+        ChipLogError(DataManagement, "Client is null");
+    }
+}
+
+CHIP_ERROR ReadClient::Resubscribe()
+{
+    mReadPrepareParams.mpResubscribeDelegate->Run(mNumRetries, mIntervalMsec, mReadPrepareParams.mShouldResubscribe);
+    if (!mReadPrepareParams.mShouldResubscribe)
+    {
+        ChipLogProgress(DataManagement, "Resubscribe has been stopped");
+        return CHIP_NO_ERROR;
+    }
+    CHIP_ERROR err = InteractionModelEngine::GetInstance()->GetExchangeManager()->GetSessionManager()->SystemLayer()->StartTimer(
+        System::Clock::Milliseconds32(mIntervalMsec), OnResubscribeTimerCallback, this);
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogProgress(DataManagement, "Fail to resubscribe with error %s", ErrorStr(err));
+        mReadPrepareParams.mShouldResubscribe = false;
```

Comment:
I think it'd odd to clear out this field, since this params struct is a snapshot of the user's original intentions, and altering it after we've received it feels wrong.

More importantly, it's not necessary.

Reply:
updated. thanks